File: src/observer/observer.py
# Observer pattern

# To run this code, Python 3.10 or above is required.

from typing import Protocol
from abc import abstractmethod
from typing_extensions import Self

# Observer abstract class
class Observer(Protocol):

    @abstractmethod
    def update(self: Self) -> None:
        pass


# Provider abstract class
class Observable(Protocol):
    observers: list[Observer]
    
    @abstractmethod
    def register(self: Self, obs: Observer) -> None:
        pass
    
    @abstractmethod
    def remove(self: Self, obs: Observer) -> None:
        pass
    
    def notify(self: Self) -> None:
        for i in self.observers:
            i.update()


# Provider class
class WeatherData(Observable):

    temperature = None
    humidity = None
    pressure = None

    def __init__(self):
        self.observers = []
    
    def register(self, obs):
        self.observers.append(obs)

    def remove(self, obs):
        self.observers.remove(obs)

    def set_data(self: Self, temp: float, hum: float, pres: float):
        self.temperature = temp
        self.humidity = hum
        self.pressure = pres
        self.notify()


# Observer class
class WeatherStation(Observer):
    
    temperature = None
    humidity = None

    def __init__(self, weatherdata):
        self.weatherdata = weatherdata
        self.weatherdata.register(self)

    def update(self):
        self.temperature = self.weatherdata.temperature
        self.humidity = self.weatherdata.humidity
        self.display()
    
    def display(self):
        print('Temperature: ', self.temperature)
        print('Humidity: ', self.humidity, end='\n\n')


# Main
data = WeatherData()

subscriber = WeatherStation(weatherdata=data)
data.set_data(temp=30.6, hum=80, pres=1013.3)
data.set_data(temp=29.8, hum=85, pres=1014.2)

File: src/observer/observer.cs
// Observer pattern

// To run this code, .NET 5 or later is required.

// Provider interface
public interface ISubject
{
    public void Register(ISubscriber subscriber);
    public void Remove(ISubscriber subscriber);
    public void Notify();
}

// Provider class
public sealed class WeatherData : ISubject
{
    float _temperature;
    float _humidity;
    float _pressure;

    public float Temperature { get { return _temperature; } }
    public float Humidity { get { return _humidity; } }
    public float Pressure { get { return _pressure; } }

    List<ISubscriber> _subscribers;

    public WeatherData()
    {
        _subscribers = new List<ISubscriber>();
    }

    public void Register(ISubscriber subscriber)
    {
        _subscribers.Add(subscriber);
    }

    public void Remove(ISubscriber subscriber)
    {
        _subscribers.Remove(subscriber);
    }

    public void Notify()
    {
        foreach (var subscriber in _subscribers)
            subscriber.Update();
    }

    public void SetData(float temperature, float humidity, float pressure)
    {
        _temperature = temperature;
        _humidity = humidity;
        _pressure = pressure;
        Notify();
    }
}

// Observer interface
public interface ISubscriber
{
    public void Update();
}

// Display feature interface
public interface IDisplay
{
    public void Display();
}

// Observer class
public sealed class Subscriber : ISubscriber, IDisplay
{
    public WeatherData WeatherData { init; get; }
    private float _temperature;
    private float _humidity;

    public Subscriber(WeatherData weatherData)
    {
        WeatherData = weatherData;
        WeatherData.Register(this);
    }

    public void Update()
    {
        _temperature = WeatherData.Temperature;
        _humidity = WeatherData.Humidity;
        Display();
    }

    public void Display()
    {
        Console.WriteLine($"Current status: Temperature {_temperature}, Humidity: {_humidity}");
    }
}

public static class Program
{
    public static void Main(string[] args)
    {
        var weatherData = new WeatherData();
        var subscriber = new Subscriber(weatherData);

        weatherData.SetData(120.0f, 40.4f, 30.4f);
        weatherData.SetData(118.0f, 40.2f, 30.3f);
        weatherData.SetData(121.0f, 40.1f, 30.2f);
    }
}
